src/main/java/core/basesyntax/Car.java

```diff
@@ -1,13 +1,87 @@
 package core.basesyntax;
 
+import java.util.ArrayList;
 import java.util.List;
+import java.util.Objects;
 
 /**
  * Make this class immutable. See requirements in the README file
  */
-public class Car {
-    private int year;
-    private String color;
-    private List<Wheel> wheels;
-    private Engine engine;
+public final class Car {
+    private final int year;
+    private final String color;
+    private final List<Wheel> wheels;
+    private final Engine engine;
+
+    public Car(int year, String color, List<Wheel> wheels, Engine engine) {
+        this.year = year;
+        this.color = color;
+        this.wheels = wheelsDeepCopy(wheels);
+        this.engine = (engine == null) ? null : engine.clone();
+    }
+
+    public int getYear() {
+        return year;
+    }
+
+    public String getColor() {
+        return color;
+    }
+
+    public List<Wheel> getWheels() {
+        return new ArrayList<>(wheels);
+    }
+
+    public Engine getEngine() {
+
+        return (engine == null) ? null : engine.clone();
+    }
+
+    public Car changeEngine(Engine engine) {
+        return new Car(year, color, wheels, engine);
+    }
+
+    public Car changeColor(String newColor) {
+        return new Car(year, newColor, wheels, engine);
+    }
+
+    public Car addWheel(Wheel newWheel) {
+        List<Wheel> temp = new ArrayList<>(wheels);
+        temp.add(newWheel.clone());
+        return new Car(year, color, temp, engine);
```

Comment:
```suggestion
         Car car = new Car(year, color, wheels, engine);
         car.wheels.add(wheel.clone());
         return car
```
In your way, you are cloning the list twice